atlasdb-commons/src/main/java/com/palantir/util/TimedRunner.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.util;
+
+import com.palantir.common.concurrent.PTExecutors;
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+import java.time.Duration;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import org.immutables.value.Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class TimedRunner {
+    private static final Logger log = LoggerFactory.getLogger(TimedRunner.class);
+
+    private final ExecutorService executor;
+    private final Duration timeoutDuration;
+
+    private TimedRunner(ExecutorService executor, Duration timeoutDuration) {
+        this.executor = executor;
+        this.timeoutDuration = timeoutDuration;
+    }
+
+    public static TimedRunner create(Duration timeoutDuration) {
+        return new TimedRunner(PTExecutors.newCachedThreadPool("timed-runner"), timeoutDuration);
+    }
+
+    public <T> T run(TaskContext<T> taskContext) {
+        Future<T> future = executor.submit(() -> taskContext.task().call());
+        final Throwable failure;
+        try {
+            return future.get(timeoutDuration.toMillis(), TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            failure = e;
+        } catch (ExecutionException e) {
+            failure = e.getCause();
+        } catch (TimeoutException e) {
+            future.cancel(true);
+            failure = e;
+        }
+
+        SafeRuntimeException exception = new SafeRuntimeException(failure);
+        try {
+            taskContext.taskFailureHandler().run();
+        } catch (Throwable t) {
+            log.warn("Shutdown failure handler threw an exception itself!", t);
+            exception.addSuppressed(t);
+        }
+        throw exception;
+    }
+
+    @Value.Immutable
+    public interface TaskContext<T> {
+        /**
+         * Executed to shut down an object.
+         */
+        Callable<T> task();
+
+        /**
+         * Executed only if the shutdownCallback times out or otherwise has an exception.
+         */
+        Runnable taskFailureHandler();
```

Comment:
nit: javadocs here need updating (it's now just a task as opposed to shutdownCallback, and the task is general I think)